chapter/8/big-data.md

```diff
@@ -1,11 +1,229 @@
+
+`JJ: what about other refinement: `
+
+**Performance**  
+In the paper, the authors measure the performance of MapReduce on two computations running on a large cluster of machines. One computation *grep* through approximately 1TB of data. The other computation *sort* approximately 1TB of data. Both computations take in the order of a hundred seconds. In addition, the backup tasks do help largely reduce execution time. In the experiment where 200 out of 1746 tasks were intentionally killed, the scheduler was able to recover quickly and finish the whole computation for just a 5% increased time.  
+Overall, the performance is very good for conceptually unrelated computations.
+
+
+## Iterative processing in Map Reduce
+
+Many a analytics workloads like K-means, logistic regression, graph processing applications like PageRank, shortest path using parallel breadth first search require multiple stages of map reduce jobs. In regular map reduce framework like Hadoop, this requires the developer to manually handle the iterations in the driver code. At every iteration, the result of each stage T is written to HDFS and loaded back again at stage T+1 causing a performance bottleneck. The reason being wastage of network bandwidth, CPU resources and mainly the disk I/O operations which are inherently slow. In order to address such challenges in iterative workloads on map reduce, frameworks like Haloop, Twister and iMapReduce adopt special techniques like caching the data between iterations and keeping the mapper and reducer alive across the iterations.
+
+**Haloop** : HaLoop: Efficient Iterative Data Processing on Large Clusters.
+
+**iMapReduce**: iMapReduce: A Distributed Computing Framework for Iterative Computation
+
+**Twister** :  Twister: a runtime for iterative MapReduce.
+
+## Map Reduce inspired large scale data processing systems
+
+**Dryad/DryadLinq** : 
+
+**Spark (big one)** : 
+
+## Declarative interfaces for the Map Reduce framework
+Map reduce provides only two high level primitives - map and reduce; that the programmers have to worry about. Map reduce takes care of all the processing over a cluster, failure and recovery, data partitioning etc. However, the framework still suffers from rigidity with respect to its one-input data format (key/value pair) and two-stage data flow. Several important patterns like joins (which could be highly complex depending on the data) are extremely hard to implement and reason about for a programmer. Sometimes the code could be become repetitive  when the programmer wants to implement most common operations like projection, filtering etc.
+Non-programmers like data scientists would highly prefer SQL like interface over a cumbersome and rigid framework. Such a high level declarative language can easily express their task while leaving all of the execution optimization details to the backend engine. Also, these kind of abstractions provide ample opportunities for query optimizations.
```

Comment:
> Non-programmers like data scientists would highly prefer SQL like interface over a cumbersome and rigid framework.

You should not make claims like these without citations or stronger argumentation. It's too hand-wavy as-is.